README.rst

```diff
@@ -0,0 +1,806 @@
+  - Each `namespace` segment must be a percent-encoded string
+  - When percent-decoded, a segment:
+
+    - must not contain a '/'
+    - must not be empty
+
+  - A URL host or Authority must NOT be used as a `namespace`. Use instead a
+    `repository_url` qualifier. Note however that for some types, the
+    `namespace` may look like a host.
+
+
+- **name**:
+
+  - The `name` is prefixed by a '/' separator when the `namespace` is not empty
+  - This '/' is not part of the `name`
+  - A `name` must be a percent-encoded string
+
+
+- **version**:
+
+  - The `version` is prefixed by a '@' separator when not empty
+  - This '@' is not part of the `version`
+  - A `version` must be a percent-encoded string
+
+  - A `version` is a plain and opaque string. Some package `type` use versioning
+    conventions such as semver for NPMs or nevra conventions for RPMS. A `type`
+    may define a procedure to compare and sort versions, but there is no
+    reliable and uniform way to do such comparison consistently.
+
+
+- **qualifiers**:
+
+  - The `qualifiers` string is prefixed by a '?' separator when not empty
+  - This '?' is not part of the `qualifiers`
+  - This is a query string composed of zero or more `key=value` pairs each
+    separated by a '&' ampersand. A `key` and `value` are separated by the equal
+    '=' character
+  - These '&' are not part of the `key=value` pairs.
+  - `key` must be unique within the keys of the `qualifiers` string
+  - `value` cannot be an empty string: a `key=value` pair with an empty `value`
+    is the same as no key/value at all for this key
+  - For each pair of `key` = `value`:
+
+    - The `key` must be composed only of ASCII letters and numbers, '.', '-' and
+      '_' (period, dash and underscore)
+    - A `key` cannot start with a number
+    - A `key` must NOT be percent-encoded
+    - A `key` is case insensitive. The canonical form is lowercase
+    - A `key` cannot contains spaces
+    - A `value` must be must be a percent-encoded string
+    - The '=' separator is neither part of the `key` nor of the `value`
+
+
+- **subpath**:
+
+  - The `subpath` string is prefixed by a '#' separator when not empty
+  - This '#' is not part of the `subpath`
+  - The `subpath` contains zero or more segments, separated by slash '/'
+  - Leading and trailing slashes '/' are not significant and should be stripped
+    in the canonical form
+  - Each `subpath` segment must be a percent-encoded string
+  - When percent-decoded, a segment:
+
+    - must not contain a '/'
+    - must not be any of '..' or '.'
+    - must not be empty
+
+  - The `subpath` must be interpreted as relative to the root of the package
+
+
+Character encoding
+~~~~~~~~~~~~~~~~~~
+
+For clarity and simplicity a `purl` is always an ASCII string. To ensure that
+there is no ambiguity when parsing a `purl`, separator characters and non-ASCII
+characters must be UTF-encoded and then percent-encoded as defined at::
+
```

Comment:
I'm not sure this is the way to go.  URLs can have UTF-8 characters, and they've benefitted countries with different character sets, e.g. Japan and China.  I think you should allow UTF-8 to be inclusive.  What if there are packages whose users don't ever write their names in roman characters?  ASCII seems very limiting there.  [See here](https://www.w3.org/International/articles/idn-and-iri/)

Reply:
I am with you there: the whole point is to avoid any weird thing and ensure that things are properly UTF-8 and then percent encoded. This means that eventually any character may be used, but using non-ASCII would require a bit more encoding and decoding.